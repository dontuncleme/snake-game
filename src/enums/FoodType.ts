export enum FoodType {
    CHERRY = "cherry",
    MUSHROOM = "mushroom",
    PIZZA = "pizza",
}
