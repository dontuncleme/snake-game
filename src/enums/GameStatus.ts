export enum GameState {
    NEW_GAME = "newGame",
    ACTIVE = "active",
    PAUSED = "paused",
    GAME_OVER = "gameOver",
}
