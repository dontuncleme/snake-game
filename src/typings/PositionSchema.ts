export interface PositionSchema {
    x: number;
    y: number;
}
