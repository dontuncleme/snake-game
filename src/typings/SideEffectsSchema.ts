import { SideEffect } from "../consts/SideEffect";

export interface SideEffectsSchema {
    [SideEffect.IS_INVERTED_CONTROLS]: boolean;
    [SideEffect.SPEED]: number;
}
