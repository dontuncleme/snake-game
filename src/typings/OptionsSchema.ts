import { GameState } from "../enums/GameStatus";

export interface OptionsSchema {
    score: number;
    bestScore: number;
    gameState: GameState;
}
