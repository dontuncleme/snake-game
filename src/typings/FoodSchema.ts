import { FoodType } from "../enums/FoodType";
import { SideEffect } from "../consts/SideEffect";
import { PositionSchema } from "./PositionSchema";

export interface FoodSchema {
    type: FoodType;
    icon: string;
    points: number;
    position: PositionSchema;
    sideEffect?: SideEffect;
}
