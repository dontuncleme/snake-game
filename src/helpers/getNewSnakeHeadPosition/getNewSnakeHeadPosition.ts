import { Direction } from "../../enums/Direction";
import { PositionSchema } from "../../typings/PositionSchema";

export function getNewSnakeHeadPosition(
    snake: PositionSchema[],
    direction: Direction,
) {
    const snakeHead = { ...snake[0] };

    // Depending on the direction, it updates the x or y coordinate of the snake's head
    switch (direction) {
        case Direction.UP:
            snakeHead.y -= 1;
            break;
        case Direction.DOWN:
            snakeHead.y += 1;
            break;
        case Direction.LEFT:
            snakeHead.x -= 1;
            break;
        case Direction.RIGHT:
            snakeHead.x += 1;
            break;
    }

    return snakeHead;
}
