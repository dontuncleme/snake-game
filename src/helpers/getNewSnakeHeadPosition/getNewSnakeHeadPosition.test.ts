import { describe, expect, it } from "vitest";
import { getNewSnakeHeadPosition } from "./getNewSnakeHeadPosition";
import { Direction } from "../../enums/Direction";
import { PositionSchema } from "../../typings/PositionSchema";

describe("getNewSnakeHeadPosition", () => {
    it.each(Object.values(Direction))(
        "should move the snake's head in the correct direction",
        (direction) => {
            const snake: PositionSchema[] = [{ x: 5, y: 5 }];
            const snakeHead = getNewSnakeHeadPosition(snake, direction);

            expect(snakeHead).toMatchSnapshot();
        },
    );
});
