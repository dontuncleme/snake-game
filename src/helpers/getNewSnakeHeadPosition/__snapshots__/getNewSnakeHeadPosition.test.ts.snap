// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getNewSnakeHeadPosition > should move the snake's head in the correct direction 1`] = `
{
  "x": 5,
  "y": 4,
}
`;

exports[`getNewSnakeHeadPosition > should move the snake's head in the correct direction 2`] = `
{
  "x": 5,
  "y": 6,
}
`;

exports[`getNewSnakeHeadPosition > should move the snake's head in the correct direction 3`] = `
{
  "x": 4,
  "y": 5,
}
`;

exports[`getNewSnakeHeadPosition > should move the snake's head in the correct direction 4`] = `
{
  "x": 6,
  "y": 5,
}
`;
